### PR TITLE
[ADP-3039] Remove WalletKey class

### PR DESCRIPTION
**Pull request by paolino**
ADP-3039

### CODE REVIEW

File: lib/wallet/src/Cardano/Wallet/Address/Keys/WalletKey.hs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeFamilies #-}
+
+
+-- ghc , at least 8.10.7 cannot figure out the constraint is necessary in
+-- liftRawKey, so we disable the warning.
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE RankNTypes #-}
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Address/Keys/WalletKey.hs]:
```suggestion
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE TypeFamilies #-}

-- ghc , at least 8.10.7 cannot figure out the constraint is necessary in
-- liftRawKey, so we disable the warning.
{-# OPTIONS_GHC -Wno-redundant-constraints #-}
```

formatting

File: lib/wallet/src/Cardano/Wallet/Address/Keys/SequentialAny.hs [hunk 2]
```diff
@@ -0,0 +1,77 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE RankNTypes #-}
+module Cardano.Wallet.Address.Keys.SequentialAny
+    where
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Address/Keys/SequentialAny.hs]:
```suggestion
module Cardano.Wallet.Address.Keys.SequentialAny
    ( mkSeqAnyState
    , mkSeqStateFromRootXPrv
    ) where
```

namespaces: export list

File: lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+-- An implementation of shared script state using
+-- scheme specified in CIP-1854 Multi-signature Wallets.
+
+module Cardano.Wallet.Address.Keys.Shared
+    (
+
+    -- mkSharedStateFromRootXPrv
+    -- , validateScriptTemplates
+
+    mkSharedStateFromAccountXPub, mkSharedStateFromRootXPrv, addCosignerAccXPub, validateScriptTemplates, toSharedWalletId) where
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Address/Keys/Shared.hs]:
```suggestion
module Cardano.Wallet.Address.Keys.Shared
    ( mkSharedStateFromAccountXPub
    , mkSharedStateFromRootXPrv
    , addCosignerAccXPub
    , validateScriptTemplates
    , toSharedWalletId
    ) where
```

formatting: export list

File: lib/wallet/src/Cardano/Wallet/Read/NetworkId.hs [hunk 4]
```diff
@@ -59,7 +59,7 @@ import qualified Data.Text as T
 --              addresses. Genesis file needs to be passed explicitly when
 --              starting the application.
 --
-data NetworkDiscriminant = Mainnet | Testnet Nat
+data NetworkDiscriminant = Mainnet | Testnet Nat | NoNetwork
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Read/NetworkId.hs]:
```suggestion
data NetworkDiscriminant = Mainnet | Testnet Nat
```

No `NoNetwork` please — we don't need it, and it's not as conceptually tight.

File: lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs [hunk 5]
```diff
@@ -541,13 +542,13 @@ signTransaction
     mkTxInWitness i = do
         addr <- resolveInput i
         (k, pwd) <- resolveAddress addr
+        let  pk = (getRawKey keyF k, pwd)
         pure $ case txWitnessTagFor @k of
-            TxWitnessShelleyUTxO ->
-                mkShelleyWitness body (getRawKey k, pwd)
+            TxWitnessShelleyUTxO -> mkShelleyWitness body pk
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs:547]:
```suggestion
            TxWitnessShelleyUTxO ->
                mkShelleyWitness body pk
```

formatting

File: lib/wallet/test/unit/Cardano/Wallet/DB/Arbitrary.hs [hunk 6]
```diff
@@ -232,6 +228,8 @@ import qualified Cardano.Wallet.Address.Derivation.Shared as Shared
 import qualified Cardano.Wallet.Address.Derivation.Shelley as Shelley
 import qualified Cardano.Wallet.Address.Discovery.Sequential as Seq
 import qualified Cardano.Wallet.Address.Discovery.Shared as Shared
+import Cardano.Wallet.Address.Keys.WalletKey
+    ( getRawKey, liftRawKey, publicKey )
```
Comment (HeinrichApfelmus) [inline lib/wallet/test/unit/Cardano/Wallet/DB/Arbitrary.hs:232]: formatting: don't mix explicit imports and qualified imports

**Review by paweljakubas — APPROVED** [review]
quite breath-taking refactoring and code reorganization and reshuffling. Went commit by commit and have not encountered anything suspicious. overall idea and rationale behind it looks fine to me. If it compiles and all tests are green then it should be fine according to my best knowledge

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 7]
```diff
@@ -2092,12 +2099,12 @@ signTransaction
...
         , Bounded (Index (AddressIndexDerivationType k) (AddressCredential k))
-        , WalletKey k
         , IsOwned s k ktype
         , HardDerivation k
-        , k ~ KeyOf s
+        , WalletFlavor s
+        , KeyOf s ~ k
         , AccountIxForStaking s
-        , ToWitnessCountCtx s
+        -- , Including '[ 'ShelleyF, IcarusF, 'SharedF] (FlavorOf s)
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
        -- , Including '[ 'ShelleyF, IcarusF, 'SharedF] (FlavorOf s)
```

Keep or delete this comment? 🤔 I think it refers to the "actual polymorphism" of this function.

File: lib/wallet/src/Cardano/Wallet/DB/Layer.hs [hunk 8]
```diff
@@ -305,10 +302,11 @@ newDBFactory tr defaultFieldValues ti = \case
                     let trDel = contramap (MsgRemovingDatabaseFile widp) tr
                     deleteSqliteDatabase trDel (databaseFile wid)
             , listDatabases =
-                findDatabases @(KeyOf s) tr databaseDir
+                findDatabases key tr databaseDir
             }
       where
-        databaseFilePrefix = keyTypeDescriptor $ Proxy @(KeyOf s)
+        key = keyOfWallet wf
+        databaseFilePrefix = keyTypeDescriptor key
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/DB/Layer.hs:309]: 😍

File: lib/wallet/cardano-wallet.cabal [hunk 9]
```diff
@@ -198,6 +198,7 @@ library
     , tracer-transformers
     , transformers
     , transformers-base
+    , type-level-sets
```
Comment (HeinrichApfelmus) [inline lib/wallet/cardano-wallet.cabal:201]:
```suggestion
```

It looks like you didn't end up needing it for the type-level stuff.

File: lib/wallet/src/Cardano/Wallet/Flavor.hs [hunk 10]
```diff
@@ -53,8 +72,39 @@ data WalletFlavorS s where
...
+     , 'BenchByronF
+     , 'BenchShelleyF
+     , 'TestStateF
+     ]
+
+type IncludingStates xs x = Including AllFlavors xs x
+
 -- | A function to reify the flavor of a state.
-class WalletFlavor s where
+class  WalletFlavor s where
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Flavor.hs]:
```suggestion
class WalletFlavor s where
```

formatting

File: lib/wallet/src/Cardano/Wallet/Flavor.hs [hunk 11]
```diff
@@ -111,6 +178,42 @@ keyOfWallet TestStateS = ShelleyKeyS
...
+notByronKey :: KeyFlavorS k
+    -> (Excluding '[ByronKey] k  => KeyFlavorS k -> x)
+    -> Maybe x
+notByronKey x h = case x of
+    ByronKeyS -> Nothing
+    ShelleyKeyS -> Just $ h ShelleyKeyS
+    IcarusKeyS -> Just $ h IcarusKeyS
+    SharedKeyS -> Just $ h SharedKeyS
+
+-- | Helper to treat
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Flavor.hs]:
```suggestion
-- | Helper lemma to specialize on a subset of wallet flavors.
```

Not sure if the "lemma" is a good word here, but I want to highlight that the main purpose of this helper function is to carry a proof.

**Review by HeinrichApfelmus — APPROVED** [review]
Very good, thank you! 😊 A couple of minor comments, one piece of superfluous code. Otherwise, happy to merge.

----

Review assertion: I have checked that the source code of each function that has been moved has not changed semantics.

**Pull request merged**